File: Ex1010_ShortestRoad/ShortestRoad.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <cfloat>

using namespace std;

typedef struct coordinate
{
	double x;
	double y;
}ElementType;

typedef double WeightType;

class Graph
{
private:
	const int vexnum;
	int arcnum;
	vector<ElementType> vertexes;
	vector<vector<WeightType> > arcs;

public:
	constexpr static double INF = DBL_MAX;

	Graph(int);
	~Graph();
	
	void AddVertex(int, ElementType);
	void AddArc(int, int);
	void DelArc(int, int);
	void PrintAdjMat(void) const;
	double Euclidean(int, int) const;
	WeightType ShortestRoad(int) const;
};

constexpr double Graph::INF;

Graph::Graph(int m) :vexnum(m), arcnum(0)
			, vertexes(vector<ElementType>(vexnum))
			, arcs(vector<vector<WeightType> >(vexnum
						, vector<WeightType>(vexnum, INF)))
{
	for (int i = 0; i < (int)arcs.size(); i++)
	{
		arcs[i][i] = 0;
	}
}

Graph::~Graph()
{
}

void Graph::AddVertex(int m, ElementType x)
{
	// If input index begin from 1
	m--;
	
	if (m >= vexnum || m < 0)
	{
		throw "Vertex index out of range!";
	}

	vertexes[m] = x;
}

void Graph::AddArc(int m, int n)
{
	// If input index begin from 1
	m--;
	n--;
	
	if (m >= vexnum || n >= vexnum  || m < 0 || n < 0)
	{
		throw "Vertex index out of range!";
	}

	if (m == n)
	{
		throw "The arc cannot connect between the same vertex.";
	}

	if (arcs[m][n] == INF)
	{
		arcnum++;
	}

	arcs[m][n] = Euclidean(m, n);

	// For undirected graph
	// arcs[n][m] = w;
}

void Graph::DelArc(int m, int n)
{
	// If input index begin from 1
	m--;
	n--;
	
	if (m >= vexnum || n >= vexnum || m < 0 || n < 0)
	{
		throw "Vertex index out of range!";
	}

	if (m == n)
	{
		throw "Cannot delete the arc between the same vertex.";
	}

	if (arcs[m][n] != INF)
	{
		arcnum--;
	}
	arcs[m][n] = INF;

	// For undirected graph
	// arcs[n][m] = INF;
}

// Print out the adjacent matrix
void Graph::PrintAdjMat(void) const
{
	for (auto& i : arcs)
	{
		for (auto& j : i)
		{
			if (j == INF)
			{
				cout << "INF";
			}
			else
			{
				cout << j;
			}
			cout << "\t";
		}
		cout << endl;
	}
}

double Graph::Euclidean(int m, int n) const
{
	if (m >= vexnum || n >= vexnum || m < 0 || n < 0)
	{
		throw "Vertex index out of range!";
	}

	if (m == n)
	{
		throw "Cannot delete the arc between the same vertex.";
	}
	
	return sqrt((vertexes[m].x - vertexes[n].x) 
		* (vertexes[m].x - vertexes[n].x)
		+ (vertexes[m].y - vertexes[n].y)
		* (vertexes[m].y - vertexes[n].y));
}

// Given one start point, return the shortest road built
// from this point
WeightType Graph::ShortestRoad(int m_start) const
{
	// If input index start from 1
	m_start--;

	if (m_start >= vexnum || m_start < 0)
	{
		throw "Vertex index out of range!";
	}

	WeightType RoadLength = 0;

	vector<bool> known(vexnum, false);

	known[m_start] = true;

	// Minimun distance to start point
	vector<WeightType> MinDist(vexnum);
	// Length of last step to this point
	// (Length of new road needed to be built)
	vector<WeightType> LastDist(vexnum);

	for (int i = 0; i < vexnum; i++)
	{
		LastDist[i] = MinDist[i] = arcs[m_start][i];
	}

	while (true)
	{
		int v_shortest = -1;
		WeightType w_shortest = INF;

		// Get the shortest unknown vertex
		for (int i = 0; i < vexnum; i++)
		{
			if (known[i])
			{
				continue;
			}
			
			if (MinDist[i] < w_shortest)
			{
				v_shortest = i;
				w_shortest = MinDist[i];
			}
		}

		// No unknown accessible vertex
		if (v_shortest == -1)
		{
			for (bool k : known)
			{
				// Found vertex unaccessible
				if (!k)
				{
					return -1;
				}
			}

			return RoadLength;
		}

		known[v_shortest] = true;
		RoadLength += LastDist[v_shortest];

		// Update MinDist and LastDist
		for (int j = 0; j < vexnum; j++)
		{
			if (known[j])
			{
				continue;
			}
			
			if (arcs[v_shortest][j] == INF)
			{
				continue;
			}

			if (w_shortest + arcs[v_shortest][j] < MinDist[j])
			{
				MinDist[j] = w_shortest + arcs[v_shortest][j];
				LastDist[j] = arcs[v_shortest][j];
			}
		}
	}
}

int main(void)
{
	int n_vex = 0;

	vector<Graph> graphs;

	cin >> n_vex;
	
	while (n_vex)
	{
		int n_arc;
		cin >> n_arc;

		Graph G(n_vex);
		
		for (int i = 1; i <= n_vex; i++)
		{
			ElementType a;
			cin >> a.x >> a.y;

			G.AddVertex(i, a);
		}

		for (int i = 1; i <= n_arc; i++)
		{
			int m, n;
			cin >> m >> n;
			G.AddArc(m, n);
		}
		
		graphs.push_back(G);

		n_vex = 0;
		cin >> n_vex;
	}

	cout << endl;
	cout << "Output:" << endl;

	cout.setf(ios_base::fixed);
	cout.precision(2);

	for (auto& G : graphs)
	{
		WeightType RoadLength = G.ShortestRoad(1);
		if (RoadLength < 0)
		{
			cout << "NO" << endl;
		}
		else
		{
			cout << RoadLength << endl;
		}
	}

	return 0;
}

File: Ex1010_ShortestRoad/ShortestRoad.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class ShortestRoad {
    public static void main(String[] args) {
        ArrayList<GraphWithCoordinate> graphs = new ArrayList<>();

        Scanner input = new Scanner(System.in);

        while (input.hasNextInt()) {
            int vexNum = input.nextInt();
            int arcNum = input.nextInt();

            GraphWithCoordinate graph
                    = new GraphWithCoordinate(vexNum, true);

            for (int i = 1; i <= vexNum; i++) {
                Coordinate c = new Coordinate(
                        input.nextDouble(), input.nextDouble());
                graph.setVertex(i, c);
            }

            for (int i = 1; i <= arcNum; i++) {
                int m = input.nextInt();
                int n = input.nextInt();
                graph.addArc(m, n);
            }

            graphs.add(graph);
        }

        System.out.println();
        System.out.println("Output:");

        for (GraphWithCoordinate g: graphs) {
            double roadLength = g.getShortestRoad(1);
            if (roadLength < 0) {
                System.out.println("NO");
            }
            else {
                System.out.printf("%.2f\n", roadLength);
            }
        }
    }
}

class Coordinate
{
    double x;
    double y;

    public Coordinate(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double euclidDistance(Coordinate c) {
        if (c == null) {
            throw new NullPointerException();
        }

        double deltaX = this.x - c.x;
        double deltaY = this.y - c.y;
        return Math.sqrt(deltaX * deltaX + deltaY * deltaY);
    }
}

// Graph with adjacent matrix
class MyGraph<E> {
    // Number of vertexes, set when construct
    private final int vexNum;

    // Number of arcs
    private int arcNum;

    // The graph is directed or not
    private final boolean directed;

    // Data for all vertexes
    private E[] vertexes;

    // Use double as WeightType
    // Weight of arcs
    private double[][] arcs;

    // The constant to represent no arc
    public final static double INF = Double.MAX_VALUE;

    // Default to construct an undirected graph
    public MyGraph(int vexNum) {
        this(vexNum, false);
    }

    public MyGraph(int vexNum, boolean directed) {
        this.vexNum = vexNum;
        this.arcNum = 0;
        this.directed = directed;

        makeEmpty();
    }

    @SuppressWarnings("unchecked")
    public void makeEmpty() {
        vertexes = (E[])new Object[vexNum];
        arcs = new double[vexNum][vexNum];

        for (int i = 0; i < vexNum; i++) {
            Arrays.fill(arcs[i], INF);
            arcs[i][i] = 0;
        }
    }

    public int getVertexNum() {
        return vexNum;
    }

    public int getArcNum() {
        return arcNum;
    }

    public boolean isDirected() {
        return directed;
    }

    // If the given does not start from 0
    private static int modifyIndex(int index) {
        return index - 1;
    }

    // Check if the index is legal
    protected void checkIndex(int m) {
        m = modifyIndex(m);
        if (m >= vexNum || m < 0) {
            throw new IndexOutOfBoundsException(
                    "Vertex index out of range.");
        }
    }

    // Check if the indices of arc is legal
    protected void checkArcIndex(int m, int n) {
        checkIndex(m);
        checkIndex(n);

        m = modifyIndex(m);
        n = modifyIndex(n);

        if (m == n) {
            throw new IllegalArgumentException(
                    "The arc cannot connect between the same vertex.");
        }
    }

    // Set the data of vertex, return the current value
    public E setVertex(int index, E data) {
        checkIndex(index);
        index = modifyIndex(index);

        E current = vertexes[index];
        vertexes[index] = data;
        return current;
    }

    // Get the data of vertex
    public E getVertex(int index) {
        checkIndex(index);
        index = modifyIndex(index);
        return vertexes[index];
    }

    // Get weight of arc from m to n
    public double getArcWeight(int m, int n) {
        checkArcIndex(m, n);
        m = modifyIndex(m);
        n = modifyIndex(n);
        return arcs[m][n];
    }

    // Set the weight of arc from m to n, return the current value
    public double setArcWeight(int m, int n, double weight) {
        checkArcIndex(m, n);
        m = modifyIndex(m);
        n = modifyIndex(n);

        if (arcs[m][n] == INF) {
            arcNum++;
        }

        double current = arcs[m][n];

        arcs[m][n] = weight;

        // For undirected graph
        if (!isDirected()) {
            arcs[n][m] = weight;
        }

        return current;
    }

    // Remove arc from m to n
    public double removeArc(int m, int n) {
        checkArcIndex(m, n);
        m = modifyIndex(m);
        n = modifyIndex(n);

        if (arcs[m][n] != INF) {
            arcNum--;
        }

        double current = arcs[m][n];

        arcs[m][n] = INF;

        // For undirected graph
        if (!isDirected()) {
            arcs[n][m] = INF;
        }

        return current;
    }

    // Show the adjacent matrix
    public void showAdjMatrix() {
        for (double[] i : arcs){
            for (double j : i){
                if (j == INF) {
                    System.out.print("INF");
                }
                else {
                    System.out.print(j);
                }
                System.out.print("\t");
            }
            System.out.println();
        }
    }
}

class GraphWithCoordinate extends MyGraph<Coordinate> {
    public GraphWithCoordinate(int vexNum) {
        this(vexNum, false);
    }

    public GraphWithCoordinate(int vexNum, boolean directed) {
        super(vexNum, directed);
    }

    public double getDistance(int m, int n) {
        return getVertex(m).euclidDistance(getVertex(n));
    }

    public double addArc(int m, int n) {
        return super.setArcWeight(m, n, getDistance(m, n));
    }

    public double getShortestRoad(int indexStart) {
        checkIndex(indexStart);

        double roadLength = 0.0;
        boolean[] known = new boolean[getVertexNum()];

        known[indexStart - 1] = true;

        // Minimun distance to start point
        double[] minDist = new double[getVertexNum()];
        // Length of last step to this point
        // (Length of new road needed to be built)
        double[] lastDist = new double[getVertexNum()];

        for (int i = 1; i <= getVertexNum(); i++) {
            lastDist[i - 1] = minDist[i - 1]
                    = indexStart != i
                    ? getArcWeight(indexStart, i) : 0.0;
        }

        while (true) {
            int vertexShortest = -1;
            double weightShortest = INF;

            // Get the shortest unknown vertex
            for (int i = 1; i <= getVertexNum(); i++) {
                if (known[i - 1]) {
                    continue;
                }

                if (minDist[i - 1] < weightShortest) {
                    vertexShortest = i;
                    weightShortest = minDist[i - 1];
                }
            }

            // No unknown accessible vertex
            if (vertexShortest == -1) {
                for (boolean k : known) {
                    // Found vertex inaccessible
                    if (!k) {
                        return -1.0;
                    }
                }

                return roadLength;
            }

            known[vertexShortest - 1] = true;
            roadLength += lastDist[vertexShortest - 1];

            // Update minDist and lastDist
            for (int j = 1; j <= getVertexNum(); j++) {
                if (known[j - 1]) {
                    continue;
                }

                if (getArcWeight(vertexShortest, j) == INF) {
                    continue;
                }

                if (weightShortest + getArcWeight(vertexShortest, j)
                        < minDist[j - 1]) {
                    minDist[j - 1] = weightShortest
                            + getArcWeight(vertexShortest, j);
                    lastDist[j - 1] = getArcWeight(vertexShortest, j);
                }
            }
        }
    }
}
